rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -0,0 +1,558 @@
+    ...  # Existing logic.
+```
+
+
+This utility is only used in the `CallbackList` object, which already handles converting `tf.Tensor` logs to `NumPy`. User-defined `Callback`s do not have to be aware of this logic and will not need changes to support `ParameterServerStrategy`.  
+
+
+##### Epoch-level callbacks
+
+Since the workers will sync every epoch anyway, fetching the remote values incurs minimal additional overhead and so epoch-level `Callback`s can be easily supported.
+
+
+##### Batch-level callbacks
+
+##### What constitutes a `step` in `Model.fit` with `ParameterServerStrategy`?
+
+There are two mental models users might have of what constitutes a single `step` when running `Model.fit` with `ParameterServerStrategy`. Which mental model we choose has implications for how users specify `steps_per_epoch` and how we handle batch-level `Callback`s.
+
+**Option 1: A `step` is one batch on one worker**
+
+This is the mental model used in custom training loops with `ParameterServerStrategy`. With this mental model, every time `Model.train_function` is called, it schedules a single batch on a single worker. If there are `W` workers, then setting `steps_per_epoch=100` means each worker will run (approximately) `100/W` batches.
+
+**Pros:**
+
+-  Users can treat the workers like a "worker pool". E.g., they do not have to be aware of how many workers are in their cluster. They do not have to change the `steps_per_epoch` parameter when they change the number of workers. `steps_per_epoch=1000` runs 1000 training batches regardless of the number of workers allocated to the cluster.
+
+**Cons:**
+
+- There is no reasonably performant way to support batch-level `Callback`s. This is because to support batch-level `Callback`s, we would need to sync after every step. When a step is defined as a single batch on a single worker, this means that only one worker will ever be non-idle when batch-level `Callback`s are passed. We cannot even rely on the `steps_per_execution` parameter, because this parameter only controls how many batches are run within one `tf.function`, but the `tf.function` is still only run on one worker. Meaning even with a high `steps_per_execution`, batch-level `Callback`s would cause all but one worker to be idle at any given time. Because of this, if we go with this mental model, we should error out when a user attempts to pass batch-level `Callback`s to `Model.fit`
+- If a user passes a `Dataset` with `D` batches to each worker, they might expect that setting `steps_per_epoch=D` will visit every batch of the `Dataset`. However, only the first `D/W` batches will be visited with this mental model.
+
+**Option 2: A `step` is one batch on every worker**
+
+With this mental model, every time `Model.train_function` is called, it schedules one batch to execute on each worker. This means that if there are `W` workers, passing `steps_per_epoch=100` will actually run `100 * W` batches of training, with each worker seeing `100` batches.
+
```

Comment:
This option would make the total steps dependent on the number of workers, which is not desirable from users' point of view. It is at odds with the programming model provided by the CTL + PSStrategy. The proposed mental model would also break down if we support elasticity in the future. 

In this option, the name of the `steps_per_epoch` argument should be changed to `steps_per_epoch_per_worker`. 

Alternatively, you can sync every `W` steps only run `steps_per_epoch/W` steps, effectively turning it into synchronous parameter server training.